File: src/rvw/presentation.py
"""Strict, immutable repository presentation settings and parsing."""

from __future__ import annotations

import unicodedata
from typing import Annotated, Any, Literal

import yaml
from pydantic import BaseModel, ConfigDict, Field, field_validator

WORKTREE_RULE_WARNING = (
    "WARNING: .rvw rules loaded from the working tree via "
    "--allow-worktree-rules; this run is non-SoT."
)


class PresentationConfigInvalid(ValueError):
    """Presentation configuration cannot be trusted or parsed."""

    reason = "presentation_config_invalid"

    def __init__(self, detail: str) -> None:
        super().__init__(f"{self.reason}: {detail}")


class VoiceConfig(BaseModel):
    """Repository-selected reviewer voice used only for presentation."""

    model_config = ConfigDict(extra="forbid", strict=True, frozen=True, serialize_by_alias=True)

    audience: Literal["engineers", "mixed"] = "engineers"
    register_: Literal["formal", "neutral"] = Field(default="formal", alias="register")
    guidance: str | None = Field(default=None, max_length=800)
    examples: list[Annotated[str, Field(max_length=200)]] = Field(
        default_factory=list, max_length=3
    )
    allowed_terms: list[str] = Field(default_factory=list)

    @field_validator("guidance")
    @classmethod
    def _plain_multiline_guidance(cls, value: str | None) -> str | None:
        if value is not None and any(
            char != "\n" and unicodedata.category(char) in {"Cc", "Cf", "Cs", "Zl", "Zp"}
            for char in value
        ):
            raise ValueError("voice guidance must contain plain text")
        return value

    @field_validator("examples")
    @classmethod
    def _bounded_examples(cls, value: list[str]) -> list[str]:
        for example in value:
            if any(
                char != "\n" and unicodedata.category(char) in {"Cc", "Cf", "Cs", "Zl", "Zp"}
                for char in example
            ):
                raise ValueError("voice examples must contain plain text")
            if len(example) > 200:
                raise ValueError("voice examples must be at most 200 characters")
        return value

    @field_validator("allowed_terms")
    @classmethod
    def _plain_allowed_terms(cls, value: list[str]) -> list[str]:
        for term in value:
            if any(
                char != "\n" and unicodedata.category(char) in {"Cc", "Cf", "Cs", "Zl", "Zp"}
                for char in term
            ):
                raise ValueError("voice allowed terms must contain plain text")
        return value


class SynthesisConfig(BaseModel):
    """Repository-selected synthesis execution setting."""

    model_config = ConfigDict(extra="forbid", strict=True, frozen=True)

    enabled: bool = True


class PresentationConfig(BaseModel):
    """Base-ref presentation snapshot, safe to retain in strict run contracts."""

    model_config = ConfigDict(extra="forbid", strict=True, frozen=True)

    display_name: str = Field(default="rvw", min_length=1, max_length=80)
    short_name: str = Field(default="rvw", min_length=1, max_length=40)
    locale: Literal["ko", "en"] = "en"
    footer: str | None = Field(default=None, max_length=240)
    voice: VoiceConfig = Field(default_factory=VoiceConfig)
    synthesis: SynthesisConfig = Field(default_factory=SynthesisConfig)

    @field_validator("display_name", "short_name")
    @classmethod
    def _nonblank_name(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("presentation name must not be blank")
        return value

    @field_validator("display_name", "short_name", "footer")
    @classmethod
    def _plain_single_line(cls, value: str | None) -> str | None:
        if value is not None and any(
            unicodedata.category(char) in {"Cc", "Cf", "Cs", "Zl", "Zp"} for char in value
        ):
            raise ValueError(
                "presentation text must be single-line and contain no control characters"
            )
        return value


class _UniqueKeyLoader(yaml.SafeLoader):
    """Reject duplicate keys rather than permitting ambiguous configuration."""

    def construct_mapping(self, node: yaml.MappingNode, deep: bool = False) -> dict[Any, Any]:
        result: dict[Any, Any] = {}
        for key_node, value_node in node.value:
            key = self.construct_object(key_node, deep=deep)
            try:
                duplicate = key in result
            except TypeError as exc:
                raise PresentationConfigInvalid("configuration keys must be strings") from exc
            if duplicate:
                raise PresentationConfigInvalid("duplicate configuration key")
            result[key] = self.construct_object(value_node, deep=deep)
        return result


def parse_presentation_config(raw: str) -> PresentationConfig:
    try:
        return PresentationConfig.model_validate(yaml.load(raw, Loader=_UniqueKeyLoader))
    except (ValueError, yaml.YAMLError) as exc:
        if isinstance(exc, PresentationConfigInvalid):
            raise
        raise PresentationConfigInvalid(str(exc)) from exc

File: cloud/worker/src/presentation.ts
import {isAlias, parseDocument, visit} from "yaml";

export type Locale = "ko" | "en";

export interface VoiceConfig {
  audience: "engineers" | "mixed";
  register: "formal" | "neutral";
  guidance: string | null;
  examples: string[];
  allowed_terms: string[];
}

export interface SynthesisConfig {
  enabled: boolean;
}

export interface PresentationConfig {
  display_name: string;
  short_name: string;
  locale: Locale;
  footer: string | null;
  voice: VoiceConfig;
  synthesis: SynthesisConfig;
}

export function defaultPresentation(): PresentationConfig {
  return {display_name: "rvw", short_name: "rvw", locale: "en", footer: null,
    voice: {audience: "engineers", register: "formal", guidance: null, examples: [], allowed_terms: []},
    synthesis: {enabled: true}};
}

function containsDisallowedText(value: string, multiline = false): boolean {
  return [...value].some((character) => character !== "\n" || !multiline
    ? /[\p{Cc}\p{Cf}\p{Cs}\p{Zl}\p{Zp}]/u.test(character)
    : false);
}

function parseVoice(value: unknown): VoiceConfig {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new Error("presentation_config_invalid");
  }
  const fields = value as Record<string, unknown>;
  if (Object.keys(fields).some((key) => !["audience", "register", "guidance", "examples", "allowed_terms"].includes(key))) {
    throw new Error("presentation_config_invalid");
  }
  const result = {...defaultPresentation().voice, ...fields};
  if (result.audience !== "engineers" && result.audience !== "mixed") {
    throw new Error("presentation_config_invalid");
  }
  if (result.register !== "formal" && result.register !== "neutral") {
    throw new Error("presentation_config_invalid");
  }
  if (!(result.guidance === null || (typeof result.guidance === "string" &&
      [...result.guidance].length <= 800 && !containsDisallowedText(result.guidance, true)))) {
    throw new Error("presentation_config_invalid");
  }
  if (!Array.isArray(result.examples) || result.examples.length > 3 || result.examples.some(
    (example) => typeof example !== "string" || [...example].length > 200 || containsDisallowedText(example, true),
  )) throw new Error("presentation_config_invalid");
  if (!Array.isArray(result.allowed_terms) || result.allowed_terms.some(
    (term) => typeof term !== "string" || containsDisallowedText(term, true),
  )) throw new Error("presentation_config_invalid");
  return result as VoiceConfig;
}

function parseSynthesis(value: unknown): SynthesisConfig {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new Error("presentation_config_invalid");
  }
  const fields = value as Record<string, unknown>;
  if (Object.keys(fields).some((key) => key !== "enabled")) throw new Error("presentation_config_invalid");
  const result = {enabled: true, ...fields};
  if (typeof result.enabled !== "boolean") throw new Error("presentation_config_invalid");
  return result;
}

/** Strict scalar contract shared with Python; optional fields use identical defaults. */
export function parsePresentation(value: unknown): PresentationConfig {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new Error("presentation_config_invalid");
  }
  const fields = value as Record<string, unknown>;
  if (Object.keys(fields).some((key) => !["display_name", "short_name", "locale", "footer", "voice", "synthesis"].includes(key))) {
    throw new Error("presentation_config_invalid");
  }
  const result = {...defaultPresentation(), ...fields};
  for (const [key, maximum] of [["display_name", 80], ["short_name", 40], ["footer", 240]] as const) {
    const text = result[key];
    if (key === "footer" && text === null) continue;
    if (typeof text !== "string" || [...text].length > maximum ||
        (key !== "footer" && text.trim().length === 0) || containsDisallowedText(text)) {
      throw new Error("presentation_config_invalid");
    }
  }
  if (result.locale !== "en" && result.locale !== "ko") throw new Error("presentation_config_invalid");
  return {...result, voice: parseVoice(result.voice), synthesis: parseSynthesis(result.synthesis)} as PresentationConfig;
}

/** Strict YAML presentation parser shared with Python's base-ref configuration contract. */
export function parsePresentationYaml(raw: string): PresentationConfig {
  try {
    if (raw.length > 16_384) throw new Error();
    const document = parseDocument(raw, {schema: "yaml-1.1", uniqueKeys: true});
    if (document.errors.length > 0 || document.warnings.length > 0) throw new Error();
    let unsafeNode = false;
    visit(document, (_key, node) => {
      const candidate = typeof node === "object" && node !== null
        ? node as {anchor?: unknown; tag?: unknown}
        : {};
      if (isAlias(node) || candidate.anchor || candidate.tag) {
        unsafeNode = true;
      }
    });
    if (unsafeNode) throw new Error();
    const value = document.toJS({maxAliasCount: 0}) as Record<string, unknown>;
    const voice = value.voice as Record<string, unknown> | undefined;
    if (!/(?:\r?\n)$/.test(raw) && typeof voice?.guidance === "string" &&
        /(?:^|\n) +guidance:\s*\|[^\n]*(?:\n +.*)*$/.test(raw) && voice.guidance.endsWith("\n")) {
      voice.guidance = voice.guidance.slice(0, -1);
    }
    return parsePresentation(value);
  } catch {
    throw new Error("presentation_config_invalid");
  }
}
